File: src/main/scala/org/tindalos/principle/domain/expectations/ThirdParty.scala
package org.tindalos.principle.domain.expectations

import org.tindalos.principle.domain.expectations.exception.InvalidConfigurationException
import org.tindalos.principle.domain.util.ListConverter

class Barrier(var layer: String = "", var components: String = "") {
  def this() = this("", "")

  def componentList() =
    if (components.isEmpty) List()
    else components
      .filter(_ >= ' ') //remove whitespaces
      .split(",").to[List].map(_.trim)
}

case class ThirdParty(var barriers: List[Barrier] = List.empty, threshold: Int = 0) extends Thresholder {

  def setBarriers(theBarriers: java.util.List[Barrier]) = {
    if (theBarriers == null)
      throw new InvalidConfigurationException("Invalid barriers expectation!")
    barriers = ListConverter.convert(theBarriers)
  }
}

File: src/main/java/org/tindalos/principle/domain/expectations/ThirdParty.java
package org.tindalos.principle.domain.expectations;

import java.util.Collections;
import java.util.List;

/**
 * Represents third-party dependency restrictions for architectural layers.
 * Defines which external libraries are allowed in specific layers through barriers.
 * Used to enforce architectural boundaries and prevent unwanted external dependencies.
 *
 * @param barriers list of barriers defining allowed libraries per layer
 * @param violationThreshold maximum number of violations allowed before failing
 */
public record ThirdParty(List<Barrier> barriers, int violationThreshold) implements IntExpectation {

    public ThirdParty {
        barriers = Collections.unmodifiableList(barriers);
    }

}

